### PR TITLE
feat(rumqttd): Add deferred acking 

### PR DESCRIPTION
**Pull request by mnpw**

### CODE REVIEW

File: rumqttd/src/segments/mod.rs [hunk 1]
```diff
@@ -367,7 +367,8 @@ mod tests {
         let mut log = CommitLog::new(max_segment_size, 1).unwrap();
 
         for i in 0..10 {
-            log.append(random_payload(i, packet_size));
+            let out = log.append(random_payload(i, packet_size));
+            println!("{out:?}");
```
Comment (henil) [inline rumqttd/src/segments/mod.rs:371]: redundant println!

File: rumqttd/src/router/routing.rs [hunk 2]
```diff
@@ -462,12 +465,18 @@ impl Router {
...
+                        publish.clone(),
                         &mut self.datalog,
                         &mut self.notifications,
                         &mut self.connections,
                     ) {
-                        Ok(_offset) => {
+                        Ok(offset_map) => {
+                            if let QoS::AtLeastOnce = qos {
+                                let ackslog = self.ackslog.get_mut(id).unwrap();
+                                ackslog.insert_pending_acks(publish, offset_map);
```
[the remark this answers is not in the document]
Comment (mnpw) [reply rumqttd/src/router/routing.rs:476]: Yep, this change of behavior lgtm

File: rumqttd/Cargo.toml [hunk 3]
```diff
@@ -25,20 +25,21 @@ tokio-tungstenite = { version = "0.15.0", optional = true }
...
-futures-util = { version = "0.3.16", optional = true}
+futures-util = { version = "0.3.16" }
 parking_lot = "0.11.2"
 config = "0.13"
 structopt = "0.3.26"
 tracing = { version="0.1", features=["log"] }
 tracing-subscriber = { version="0.3.16", features=["env-filter"] }
 metrics = "0.20.1"
 metrics-exporter-prometheus = "0.11.0"
+disk = { git = "https://github.com/bytebeamio/uplink"}
```
Comment (tekjar) [inline rumqttd/Cargo.toml:36]: Copy this as a module

File: rumqttd/rumqttd.toml [hunk 4]
```diff
@@ -11,10 +11,12 @@ max_segment_count = 10
 max_read_len = 10240
 max_connections = 10010
 
+
 # Configuration of server and connections that it accepts
 [v4.1]
 name = "v4-1"
 listen = "0.0.0.0:1883"
+persistence = false
```
Comment (tekjar) [inline rumqttd/rumqttd.toml:19]: Make this optional

File: rumqttd/src/link/local.rs [hunk 5]
```diff
@@ -45,6 +45,7 @@ impl Link {
         clean: bool,
         last_will: Option<LastWill>,
         dynamic_filters: bool,
+        _persistent: bool,
```
Comment (tekjar) [inline rumqttd/src/link/local.rs:48]: Remove all `_persistent`s

File: rumqttd/src/link/local.rs [hunk 6]
```diff
@@ -80,12 +81,29 @@ impl Link {
...
+            dynamic_filters,
+            persistent,
+        );
+        if let Event::Connect {
+            ref mut outgoing,
+            ref mut connection,
+            ..
+        } = message
+        {
+            connection.persistent = persistent;
```
Comment (tekjar) [inline rumqttd/src/link/local.rs:103]: Double check if this flag in `Connection` and `Outgoing` are being used

File: rumqttd/src/protocol/mod.rs [hunk 7]
```diff
@@ -752,3 +752,6 @@ pub trait Protocol {
     fn read_mut(&mut self, stream: &mut BytesMut, max_size: usize) -> Result<Packet, Error>;
     fn write(&self, packet: Packet, write: &mut BytesMut) -> Result<usize, Error>;
 }
+
+pub trait AsyncProtocol: Protocol + Clone + Send + 'static {}
```
Comment (tekjar) [inline rumqttd/src/protocol/mod.rs:756]: Add these extra traits directly to `Protocol`

File: rumqttd/src/link/local.rs [hunk 8]
```diff
@@ -284,6 +309,15 @@ impl LinkTx {
         self.router_tx.try_send((self.connection_id, message))?;
         Ok(())
     }
+
+    pub async fn persist(&mut self, written: HashMap<FilterIdx, Offset>) -> Result<(), LinkError> {
```
Comment (tekjar) [inline rumqttd/src/link/local.rs:313]: Rename this to `ack`

**Pull request merged**